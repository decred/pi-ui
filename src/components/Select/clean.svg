<svg xmlns="http://www.w3.org/2000/svg" width="5.914" height="6.001" viewBox="0 0 5.914 6.001"><g transform="translate(0.357 0.351)"><line y1="5.3" x2="5.2" fill="none" stroke="#f8a085" stroke-miterlimit="10" stroke-width="1"/><line x2="5.2" y2="5.3" fill="none" stroke="#f8a085" stroke-miterlimit="10" stroke-width="1"/></g></svg>