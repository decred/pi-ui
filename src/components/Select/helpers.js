export const blankValue = { value: "", label: "" };
