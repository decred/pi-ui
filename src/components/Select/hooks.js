import { useRef, useCallback, useEffect } from "react";

export function useClickOutside(onClickHandler) {
  const ref1 = useRef();
  const ref2 = useRef();

  const handleClick = useCallback(
    (event) => {
      if (
        ref1.current &&
        !ref1.current.contains(event.target) &&
        ref2.current &&
        !ref2.current.contains(event.target)
      ) {
        onClickHandler();
      }
    },
    [onClickHandler]
  );

  useEffect(() => {
    // Bind the event listener
    document.addEventListener("mousedown", handleClick);
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener("mousedown", handleClick);
    };
  });

  const callbackRef = useCallback((node1, node2) => {
    ref1.current = node1;
    ref2.current = node2;
  }, []);

  return callbackRef;
}

export function useHandleKeyboardHook(onTypeHandler) {
  const handleKeyboard = useCallback((event) => onTypeHandler(event), [
    onTypeHandler
  ]);

  useEffect(() => {
    // Bind the event listener
    document.addEventListener("keydown", handleKeyboard);
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener("keydown", handleKeyboard);
    };
  });
}

export function handleKeyboard() {}
