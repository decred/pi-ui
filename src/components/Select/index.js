export { default as SelectWrapper } from "./Select.jsx";
export { default as Select } from "./BasicSelect/BasicSelect.jsx";
export { default as Creatable } from "./CreatableSelect/CreatableSelect.jsx";
export { default as MultiSelect } from "./MultiSelect/MultiSelect.jsx";
export { default as AsyncSelect } from "./AsyncSelect/AsyncSelect.jsx";
