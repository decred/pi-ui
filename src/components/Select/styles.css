.cancelContent {
  display: inline-block;
  height: 20px;
}

.cancelContent svg {
  fill: #aaa;
  transition: fill 0.15s ease;
}

.downArrowContent svg {
  fill: #aaa;
  transition: fill 0.15s ease;
}

.cancelContent svg:hover {
  fill: darkslategray;
}

.downArrowContent svg:hover {
  fill: darkslategray;
}

.cancelContainer {
  display: inline-block;
  position: relative;
  cursor: pointer;
  height: 20px;
}

.mySelectDefault .cancelContainer {
  display: none;
  margin: 0.5em 0.5em 0.5em 0.5em;
}

.mySelectDefault.valueSelected .cancelContainer {
  display: inline-block;
}

.cancelContainer:hover .cancelContent {
  color: lightgray;
}

.cancelContainer:active .cancelContent {
  color: lightblue;
}

.mainSectionWrapper {
  display: flex;
  align-items: center;
  width: 100%;
}

.selectedDisplay {
  display: inline-block;
  position: relative;
  margin: 0.5em 0 0.5em 0.5em;
  text-align: left;
  vertical-align: middle;
}

.mySelectDefault.fixedHeight .selectedDisplay {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.mySelectDefault .fieldset .legend {
  /*these 3 defs are totally for style. I also like having no border.*/
  border: 0px solid slategray;
  background: white;
  font-style: italic;
  color: #555;

  /*these 3 defs are what makes the legend overlap the top edge of the selection container.*/
  position: absolute;
  top: 0;
  transform: translateY(-50%);

  /*basic placement of the legend*/
  padding: 0 0.5em;
  margin-left: 0.5em;
}

.mySelectDefault.roundedCorners .fieldset .legend {
  border-radius: 0.5em;
}

.mySelectDefault .downArrowContainer {
  height: 20px;
  margin-left: auto;
  cursor: pointer;
  margin-right: 0.5em;
}

.mySelectDefault .mySelectDefaultOptions {
  --display: none;
  background-color: white; /* pure style...if you make this transparent it will not really cover up things beneath it when it pops open*/
  border: 1px solid #ccc;

  text-align: left; /*you should align your options differently if you really want to*/

  /*these 3 entries place the option container directly below the selection box*/
  position: absolute;
  top: 100%;
  left: 0;
  margin-top: 10px;
  padding: 5px 0;
  border-radius: 4px;

  width: 100%; /*option list is officially inside the "fieldset" box so this width ensures the options are the same width as the selection ctonatiner...which isn't necessary*/
  max-height: 400px; /*having some max height is not strictly necessary, but usually preferable for user experience.*/
  overflow-y: auto; /*ensures options are scrollable if there are more than height of container*/

  /* border-radius: 0 0 0.5em 0.5em; */
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19); /*Make option list appear to be a bit 3-dimensional*/

  z-index: 1000; /*ensure that container for options is on top of everything else*/
}

.mySelectDefault.selectOpen .mySelectDefaultOptions {
  display: block;
}

.mySelectDefault.selectClosed .mySelectDefaultOptions {
  display: none;
}

.mySelectDefault .mySelectDefaultOptions div {
  padding: 0.5em 0 0.5em 0.5em;
  cursor: pointer;
}

.mySelectDefault .mySelectDefaultOptions div:hover {
  background-color: #eee;
}

.mySelectDefault .mySelectDefaultOptions div.focused_option {
  background-color: #1a59f7;
  color: white;
}

.mySelectDefault .mySelectDefaultOptions div:active {
  background-color: #ddd;
}

.mySelectDefault.roundedCorners .mySelectDefaultOptions div:active {
  /* border-radius: 0.8em; */
  background-color: #d4f0fd !important;
}

.mySelectDefault.roundedCorners
.mySelectDefaultOptions
div:hover:not(.focused_option) {
  /* border-radius: 0.8em; */
  background-color: #d4f0fd !important;
}

.mySelectDefault {
  position: relative;
  margin-right: 0.5em;
  vertical-align: bottom;
}

/*TODO: Extract out variable styles into separate class: e.g. roundedCorners*/
.mySelectDefault .fieldset {
  /* position: relative; all the other pieces of this component (legend, options, selectedvaluedisplay) are all inside so we want to have their placement be relative */
  /* width: 100%; fieldset is inside the top container .mySelectDefault...so this should probably always be 100%...widths are set on the top container and can be managed with component props */

  /* border: 1px solid #ccc; */
  /* background-color: white; */
  /* margin-top: 0.55em; needed to space next items below .5em should nearly overlap with something above it */
  /* padding: 0.2em 0.2em 0.2em 0.2em; the right padding is much larger to allow for the absolute positioning of the dropdowncontainer */

  /* min-height: 2.2em; this is here to handle empty state not shrinking the field set too much. You can remove it this if you don't mind an empty select being compressed. */
  /* min-width: 50px; */

  /*these 2 make placement and alignment of selected values and the cancel button easy*/
  /* display: flex; */
  /* align-items: center; */

  /* -webkit-align-items: center; */
  -webkit-box-align: center !important;
  /* -ms-flex-align: center !important; */
  align-items: center !important;
  background-color: hsl(0, 0%, 100%) !important;
  border-color: hsl(0, 0%, 80%);
  border-radius: 4px !important;
  border-style: solid !important;
  border-width: 1px !important;
  cursor: default !important;
  /* display: -webkit-box !important; */
  /* display: -webkit-flex !important; */
  /* display: -ms-flexbox !important; */
  display: flex !important;
  /* -webkit-flex-wrap: wrap !important; */
  /* -ms-flex-wrap: wrap !important; */
  flex-wrap: wrap !important;
  -webkit-box-pack: justify !important;
  /* -webkit-justify-content: space-between !important; */
  /* -ms-flex-pack: justify !important; */
  justify-content: space-between !important;
  min-height: 38px !important;
  outline: 0 !important;
  position: relative !important;
  /* -webkit-transition: all 100ms !important; */
  transition: all 100ms !important;
  box-sizing: border-box !important;
}

.mySelectDefault.selectOpen .fieldset {
  border-color: #2684ff;
  box-shadow: 0 0 0 1px #2684ff;
}

.mySelectDefault.roundedCorners {
  border-radius: 0.5em; /*if you change this style you should also change the border radii for downArrowContainer as they need to match to make things look nice.*/
}

.mySelectDefault.roundedCorners .fieldset {
  border-radius: 0.5em;
}

.mySelectDefault.selectOpen.roundedCorners .fieldset {
  border-radius: 0.5em 0.5em 0 0;
}
