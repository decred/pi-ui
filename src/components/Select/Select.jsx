import React, { useState, useRef, useCallback } from "react";
import styles from "./styles.css";
import { useClickOutside, useHandleKeyboardHook } from "./hooks";
import PropTypes from "prop-types";
import { classNames } from "../../utils";
import cleanIcon from "./clean.svg";

const blankValue = { value: "", label: "" };

const Select = ({
  onChange,
  defaultValue,
  legendLabel,
  options,
  cancelable,

  creatable,
  isValidNewOption,
  newOptionCreator,
  promptTextCreator,

  getOptionLabel,
  getOptionValue,

  optionRenderer,
  valueRenderer
}) => {
  const optionContainerRef = useRef(null);
  const dropdownButton = useRef(null);

  const [selectOpen, setSelectOpen] = useState(false);
  const [focusedOptionIndex, setFocusedOptionIndex] = useState(0);

  const [newOption, setNewOption] = useState("");
  const [addingNewOption, setAddingNewOption] = useState(false);

  const newOptions = useRef([]);

  const getValueKey = getOptionValue || (({ value }) => value);
  const getLabelKey = getOptionLabel || (({ label }) => label);

  function getOptions() {
    if (!creatable) return options;
    const _options = [...options, ...newOptions.current].filter(
      (item, pos, self) =>
        self.findIndex((_item) => getValueKey(_item) === getValueKey(item)) ===
        pos
    );

    console.log(_options);

    _options.unshift({
      label: "Type to add a new option",
      value: ""
    });
    return _options;
  }

  const [currentOptionSelected, setCurrentOptionSelected] = useState(
    (defaultValue &&
      getOptions().find((x) => getValueKey(x) === defaultValue)) ||
      blankValue
  );

  function setOption(option, knownIndex) {
    const index =
      knownIndex ||
      getOptions().findIndex(
        (opt) =>
          getValueKey(opt) === getValueKey(option) &&
          getLabelKey(opt) === getLabelKey(option)
      );
    setCurrentOptionSelected(option);
    setFocusedOptionIndex(index);
    setFocusedOptionIndex(0);
    setSelectOpen(false);
  }

  useClickOutside(() => {
    setFocusedOptionIndex(0);
    setSelectOpen(false);
  })(optionContainerRef.current, dropdownButton.current);

  useHandleKeyboardHook((e) => {
    if (selectOpen) {
      const maxOptionIndex = getOptions().length - 1;
      if (e.key === "ArrowDown") {
        const newIndex =
          focusedOptionIndex === maxOptionIndex ? 0 : focusedOptionIndex + 1;
        optionContainerRef.current
          .querySelector(`div[index="${newIndex}"]`)
          .scrollIntoViewIfNeeded(false);
        setFocusedOptionIndex(newIndex);
      }
      if (e.key === "ArrowUp") {
        const newIndex =
          focusedOptionIndex === 0 ? maxOptionIndex : focusedOptionIndex - 1;
        optionContainerRef.current
          .querySelector(`div[index="${newIndex}"]`)
          .scrollIntoViewIfNeeded(true);
        setFocusedOptionIndex(newIndex);
      }
      if (e.key === "Enter") {
        const optionIndex = focusedOptionIndex;
        const optionByIndex = getOptions()[optionIndex];
        if (onChange) {
          onChange(e, getValueKey(optionByIndex), optionByIndex);
        }
        setOption(optionByIndex, optionIndex);
      }
    }
  });

  function findOptionWrapper(el) {
    return el.getAttribute("index") ? el : findOptionWrapper(el.parentNode);
  }

  function optionSelected(e) {
    if (creatable) {
      const optionWrapper = findOptionWrapper(e.target);
      const optionIndex = parseInt(optionWrapper.getAttribute("index"), 10);
      const optionByIndex = getOptions()[optionIndex];
      if (onChange) {
        console.log("optionByIndex", optionByIndex);

        setNewOption(getLabelKey(optionByIndex));
        onChange(e, getValueKey(optionByIndex), optionByIndex);
      }
      setOption(optionByIndex, optionIndex);
    } else {
      const optionWrapper = findOptionWrapper(e.target);
      const optionIndex = parseInt(optionWrapper.getAttribute("index"), 10);
      const optionByIndex = getOptions()[optionIndex];
      if (onChange) {
        onChange(e, getValueKey(optionByIndex), optionByIndex);
      }
      setOption(optionByIndex, optionIndex);
    }
  }

  const openSelect = useCallback(() => {
    setSelectOpen((selectOpen) => !selectOpen);
  }, []);

  function cancelSelection(e) {
    if (onChange) {
      onChange(e);
    }
    setCurrentOptionSelected(blankValue);
    setFocusedOptionIndex(0);
    setSelectOpen(false);
    e.stopPropagation();
  }

  const isValidNewOptionCallback = useCallback(
    () => isValidNewOption(newOption),
    [isValidNewOption, newOption]
  );

  const promptTextCreatorCallback = useCallback(
    () => promptTextCreator(newOption),
    [promptTextCreator, newOption]
  );

  function onCreatableChange(e) {
    console.log(e.target);

    const _newOption = e.target.value;

    console.log(e.target.value);

    setNewOption(_newOption);
    setAddingNewOption(
      !getOptions().find((option) => getLabelKey(option) === _newOption)
    );

    console.log(newOption, addingNewOption);
  }

  const newOptionCreatorCallback = () => {
    newOptions.current.push({ label: newOption, value: newOption });
    onCreatableChange({ target: { value: newOption } });
    setSelectOpen(false);
    newOptionCreator(newOption);
  };

  const coreCssClassNames = classNames(
    styles.mySelectDefault,
    getValueKey(currentOptionSelected) && styles.valueSelected,
    selectOpen ? styles.selectOpen : styles.selectClosed
  );

  return (
    <div className={coreCssClassNames}>
      <div className={styles.fieldset}>
        {legendLabel && (
          <label className={styles.textinputLabel}>{legendLabel}</label>
        )}
        <div
          className={styles.mainSectionWrapper}
          onClick={openSelect}
          ref={dropdownButton}>
          {creatable ? (
            <input
              className={styles.textinput}
              value={newOption}
              onChange={onCreatableChange}
            />
          ) : (
            <div className={styles.selectedDipromptTextCreatorsplay}>
              {currentOptionSelected !== blankValue &&
                (valueRenderer
                  ? valueRenderer(currentOptionSelected)
                  : getLabelKey(currentOptionSelected))}
            </div>
          )}
          {!cancelable ? null : (
            <div className={styles.cancelContainer} onClick={cancelSelection}>
              <img src={cleanIcon} />
            </div>
          )}
          <span className={styles.separator} />
          <div className={styles.downArrowContainer}>
            <div className={styles.downArrowContent} />
          </div>
        </div>
        <div className={styles.mySelectDefaultOptions} ref={optionContainerRef}>
          {!addingNewOption &&
            getOptions().map((opt, index) => (
              <div
                onClick={optionSelected}
                key={index}
                index={index}
                className={
                  index === focusedOptionIndex && styles.focusedOption
                }>
                {opt !== blankValue &&
                  (optionRenderer ? optionRenderer(opt) : getLabelKey(opt))}
              </div>
            ))}
          {addingNewOption && isValidNewOptionCallback(newOption) && (
            <div onClick={newOptionCreatorCallback}>
              {promptTextCreatorCallback()}
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

Select.propTypes = {
  onChange: PropTypes.func.isRequired,
  options: PropTypes.array,
  defaultValue: PropTypes.object,
  legendLabel: PropTypes.string,
  cancelable: PropTypes.bool,
  creatable: PropTypes.bool,
  isValidNewOption: PropTypes.func,
  newOptionCreator: PropTypes.func,
  promptTextCreator: PropTypes.func,
  getOptionLabel: PropTypes.func,
  getOptionValue: PropTypes.func,
  optionRenderer: PropTypes.func,
  valueRenderer: PropTypes.func
};

Select.defaultProps = {
  options: [],
  defaultValue: blankValue,
  legendLabel: "",
  cancelable: false,
  creatable: false,
  isValidNewOption: undefined,
  newOptionCreator: undefined,
  promptTextCreator: undefined,
  getOptionLabel: undefined,
  getOptionValue: undefined,
  optionRenderer: undefined,
  valueRenderer: undefined
};

export default Select;
