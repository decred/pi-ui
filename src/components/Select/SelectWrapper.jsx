import React from "react";
import { classNames } from "../../utils";
import styles from "./styles.css";
import { SelectOptions } from "./helpers";
import { animated } from "react-spring";

const SelectWrapper = (
  Loading,
  Footer,
  Input,
  Controls,
  condition,
  {
    containerRef,
    menuOpened,
    openMenu,
    transitions,

    focusedOptionIndex,
    setFocusedOptionIndex,
    _options
  },
  {
    disabled,
    clearable,
    options,
    label,
    separator,
    getOptionValue,

    getOptionLabel,
    optionRenderer,
    selectOption,

    className,
    searchable,
    value,
    inputValue,
    ...props
  }
) => {
  const parentClassNames = classNames(
    styles.select,
    (Array.isArray(value) ? value.length : getOptionValue(value)) &&
      styles.valueSelected,
    searchable && inputValue && styles.search,
    menuOpened ? styles.menuOpened : styles.menuClosed,
    separator && styles.hasSeparator,
    clearable && styles.clearable,
    disabled && styles.disabled,
    className
  );

  return (
    <div className={parentClassNames} {...props}>
      <div className={styles.fieldset} ref={containerRef}>
        {label && <label className={styles.label}>{label}</label>}
        <div className={styles.controls} onClick={openMenu}>
          {Input}
          {Controls}
        </div>
        {transitions.map(
          ({ item, key, props }) =>
            item && (
              <animated.div className={styles.menu} key={key} style={props}>
                {condition ? (
                  <SelectOptions
                    options={_options} //
                    value={value}
                    selectOption={selectOption} //
                    focusedOptionIndex={focusedOptionIndex} //
                    setFocusedOptionIndex={setFocusedOptionIndex} //
                    optionRenderer={optionRenderer} //
                    getOptionLabel={getOptionLabel} //
                    getOptionValue={getOptionValue}
                  />
                ) : (
                  Loading
                )}
              </animated.div>
            )
        )}
      </div>
      {Footer}
    </div>
  );
};

export default SelectWrapper;
