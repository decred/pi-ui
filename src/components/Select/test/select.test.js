import React from "react";
import Select from "../Select";
import { create } from "react-test-renderer";
import { render } from "@testing-library/react";
import selectEvent from "react-select-event";

describe("Select component", () => {
  test("Matches the snapshot", () => {
    const props = [
      {
        value: "top",
        label: "Top"
      }
    ];
    const select = create(<Select options={props} />);
    expect(select.toJSON()).toMatchSnapshot();
  });

  test("Call on change function on option click", async () => {
    const options = [
      {
        value: "top",
        label: "Top"
      }
    ];
    let selected;
    const mockHandleChange = jest.fn((option) => {
      selected = option.value;
    });
    const { getByLabelText } = render(
      <>
        <label htmlFor="sort">Sort</label>
        <Select options={options} onChange={mockHandleChange} inputId="sort" />
      </>
    );

    // select top option
    await selectEvent.select(getByLabelText("Sort"), ["Top"]);
    expect(mockHandleChange).toBeCalled();
    expect(selected).toBe("top");
  });
});
