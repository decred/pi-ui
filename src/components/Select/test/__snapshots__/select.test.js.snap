// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component Matches the snapshot 1`] = `
<div>
  <div
    className=" css-1n3r0gh-container"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-2-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className=" css-1jhmgce-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className=" css-319lph-ValueContainer"
      >
        <div
          className=" css-14el2xx-placeholder"
          id="react-select-2-placeholder"
        >
          Select...
        </div>
        <input
          aria-autocomplete="list"
          aria-describedby="react-select-2-placeholder"
          aria-expanded={false}
          aria-haspopup={true}
          aria-readonly={true}
          className="css-mohuvp-dummyInput-DummyInput"
          disabled={false}
          id="react-select-2-input"
          inputMode="none"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          role="combobox"
          tabIndex={0}
          value=""
        />
      </div>
      <div
        className=" css-1hb7zxy-IndicatorsContainer"
      >
        <span
          className=" css-43ykx9-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          className=" css-tlfecz-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="arrowAnchor"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
